### ISSUE TITLE
Does not reliably restore scroll position using browser/mouse nav buttons in Firefox.

### ISSUE THREAD

**Issue by benjivm** [inertiajs/inertia#1698]
### Version:

- `@inertiajs/vue3` version: `^1.0.11`

### Describe the problem:

Browser/mouse button page navigation (forward/back) does not restore scroll position. Oddly, it does appear to work fine if the content is static (not served from a controller, e.g.)

### Steps to reproduce:

Reproduction repo: https://github.com/benjivm/inertiajs-firefox-scroll-position

Open the page in Firefox (this issue does not affect Chrome).

Scroll down to about the middle of the page where the link and explanation are then navigate with your browser's nav buttons or mouse button.[

[This polyfill](https://github.com/janpaul123/delayed-scroll-restoration-polyfill) makes it more reliable but is jarring to use.

This relates to this issue #1459

**Comment by Hasan-Mir** [comment]
I have the same issue (mostly in Firefox but I think I saw it a few times in Chrome as well).

**Comment by ivalkenburg** [comment]
Having the same issue. It does not always scroll to top when navigating. Specially when further down the page. Very odd.

**Comment by fritz-c** [comment]
Inertia's popstate listener (popstate being the event called when the back button is used) will asynchronously load the previous component, rendering it when it has loaded.
https://github.com/inertiajs/inertia/blob/f34f373692a964933e4272fc49fa48552301e568/packages/core/src/router.ts#L480-L492

The problem arises in that the browser wants to restore the previous scroll position as soon as the popstate listener exits, when the url changes, and at that point in time, the previous component has not loaded yet. So what you might be seeing is your pre-back page getting scrolled down (by browser native restoration) but hitting the bottom of the scroll bar, after which point the previous component gets rendered, just at the scroll position you managed to get to from the previous page.

I looked for an elegant solution around this issue, but failing to find one, made one that at least looks good from an end-user's perspective. It's for Vue+Typescript, but could be adapted for other environments. You would want to run this only once, probably in a global layout or something.

```ts
import { onMounted, onUnmounted } from 'vue';
import { debounce } from 'lodash-es';
import { router } from '@inertiajs/vue3';

/**
 * Restores the window scroll position when navigating back to a page.
 *
 * The 'popstate' event handler in inertia loads the component from the previous
 * page asynchronously, so browser-native scroll restoration is triggered before
 * the new component is mounted, resulting in the scrollbar banging against the
 * bottom of the window in the case that the page before navigating back is
 * smaller in height than the scroll position to be restored. This composable
 * separately tracks the window scroll position, saves it in the history state
 * and restores it immediately after navigation, avoiding a flash of the
 * incompletely restored scroll position.
 */
export function useWindowScrollRestoration() {
    const saveWindowScroll = debounce((event: Event) => {
        if (event.target === document)
            router.remember(window.scrollY, 'windowScrollY');
    }, 100);
    function restoreWindowScroll() {
        const windowScrollY = router.restore('windowScrollY');
        if (typeof windowScrollY === 'number') {
            window.scrollTo(0, windowScrollY);
        }
    }

    let removeListener = () => {};
    onMounted(() => {
        // Disable browser-native scroll restoration
        window.history.scrollRestoration = 'manual';

        window.addEventListener('scroll', saveWindowScroll, true);
        removeListener = router.on('navigate', restoreWindowScroll);
    });
    onUnmounted(() => {
        window.history.scrollRestoration = 'auto';
        window.removeEventListener('scroll', saveWindowScroll, true);
        removeListener();
    });
}
```

**Comment by pedroborges** [comment]
@benjivm @Hasan-Mir @ivalkenburg @fritz-c please test if #1980 fixes this issue for you.

**Comment by benjivm** [comment]
@pedroborges Hello, how do I install the version with that PR merged?

**Comment by pedroborges** [comment]
@benjivm it's not trivial because Inertia.js uses a monorepo, here's how I did it on one of my projects:

```
npm i -D https://gitpkg.now.sh/inertiajs/inertia/packages/core?master
cd node_modules/@inertiajs/core
npm i
npm run build
rm -rf node_modules

npm i -D https://gitpkg.now.sh/inertiajs/inertia/packages/svelte?master
cd node_modules/@inertiajs/svelte
npm i
npm run build
rm -rf node_modules
```

v1.3 will be tagged next week 🤞

**Comment by benjivm** [comment]
I'm using Vue and so changed the 2nd group of commands to the Vue3 package, and unfortunately still see the issue.

**Comment by Rubrasum** [comment]
@benjivm I am also using vue3. After trying for a while I found I probably needed to go the long way around, through the contribution guide and clone it, then setup a playground for my app. I took a noob-shot at setting up through a basic package.json change could not get it. 

### I also have this problem and will check on the release that is fixed. Here is summary of my problem: 

I could not get anchor links to work AND consistent scroll position on back/forward navigation. The anchor link problem was fixed by removing smooth scrolling. 

**EXAMPLE:**
If I click an anchor link to another page on my basic lead generation site:
```
<Link :href="file.href" class="w-full relative"> ... Link to other page and section ... </Link>
```

 I am automatically scrolled to the element (so long as I do not use smooth scrolling). If I use smooth scrolling it scrolls to the wrong location (ok, fine, not a big deal) . But if I press back in either case, the correct page loads at the incorrect scroll position. The position is usually lower than expected.  
 
Here's an example of what I am talking about: https://youtu.be/rfQxhHv5Szs

I will check back at the new release to see if this fixes it. I'm nearing the end of a project so I'd rather not share the repo right now, hope that's ok. 

Here's my current versions though:

npm (relevant dependencies and devdependencies together)

```
        "@inertiajs/inertia": "^0.11.1",
        "@inertiajs/progress": "^0.2.7",
        "@inertiajs/server": "^0.1.0",
        "@inertiajs/vue3": "^1.2.0",
        "laravel-vite-plugin": "^1.0",
        "vite": "^5.0",
        "ziggy-js": "^2.3.0",
        "@vitejs/plugin-vue": "^5.0.0",
```
composer versions
```
        "laravel/sail": "^1.29",
        "php": "^8.2",
        "inertiajs/inertia-laravel": "^1.0",
        "laravel/framework": "^11.0",
        "laravel/jetstream": "^5.1",
        "laravel/sanctum": "^4.0",
        "laravel/tinker": "^2.9",
        "tightenco/ziggy": "^2.3"

```

**Comment by rodrigopedra** [comment]
@Rubrasum how did you remove smooth scrolling?

**Comment by Rubrasum** [comment]
> ~@Rubrasum how did you remove smooth scrolling?~

> Although this is not a fix _per se_, as this options is toggled by default on new Firefox installs, maybe it can guide to better understand how to work around it, or to report a bug to Firefox.

It was part of my own css. I'm using Laravel, Tailwind, and Inertia Vue3. So I have an app.css file here `resources/monumental_theme/css/app.css` with these line at the top:

```
@tailwind base;
@tailwind components;
@tailwind utilities;

html {
    scroll-behavior: smooth;
}
```

So do a search for "scroll-behavior: smooth" Or Tailwind's "scroll-smooth" class. I think it being on the html tag is relevant.

**Comment by rodrigopedra** [comment]
In my case, I didn't have any string "smooth" on my generated CSS. Then I remembered this Firefox's setting:

![Screenshot_20240924_190700](https://github.com/user-attachments/assets/8ba73d0f-4ba6-4be9-b63c-3b5267e70261)

Toggling it off did the trick for me.

**Comment by Rubrasum** [comment]
Lol, great another bug you found for me too then. 

Instead of removing the smooth line, I should've added auto. If you specifically add auto it will override the default firefox setting. I just tested it very haphazardly and seemed to work. 

- Went to localhost on Firefox
- Tested the link, and it smooth scrolled badly. 
- Updated with...
```
html {
    scroll-behavior: auto;
}
```
- and it worked (jumped straight to the element, effective but not smooth).

**Comment by rodrigopedra** [comment]
I'm out of office right now, tomorrow I'll give it a try and let you know.

thanks!

**Comment by rodrigopedra** [comment]
Hey @Rubrasum, that seems to fix it! Thanks a lot!

```css
html {
    scroll-behavior: auto;
}
```

**Comment by benjivm** [comment]
Inertia [v2.0.0-beta.1](https://github.com/inertiajs/inertia/releases/tag/v2.0.0-beta.1) appears to [resolve this issue](https://github.com/benjivm/inertiajs-firefox-scroll-position) for me without resorting to the `scroll-behavior` CSS mentioned above.

**Comment by Rubrasum** [comment]
Hi #1980 seemed to resolve all the unexpected behavior I had in navigation. 

I still had a problem specific to me that I thought I'd document here and maybe someone else thinks it's issue-worthy. 

I have this in my header as an animated transition. Basic class change animation with a ref. It relies on a watcher to check the url and update the ref. The point is to create a shorter header on pages that are not the home page. 
```
       <div
            class="relative w-full overflow-hidden z-0 items-center transition-all duration-1000 ease-in-out"
            :class="isHomePage ? 'h-[44rem] tablet-sm:h-[48rem] duration-1000' : 'h-24 tablet-sm:h-60  laptop-sm:h-72'"
        >
```

Even after #1980 , if I scroll down somewhere on the home page, click a link, and try to use back, it causes jitters and is off by a bit. I have determined this animation as the cause. Funny thing is it doesn't cause any problems when shrinking the header. 

I have resolved this by turning off the animation when clicking a link TO the home page. And could probably refine it further by setting it to only have the animation turned off when going to a scroll position on the home page specifically. 

In summary, if you have an animation in your layout/header that affects the content height for the entire page, this will blow your scroll position and cause jitters so long as the animation is running. IF you turn off the animation, it will work fine even if the header/layout heights are different

This shows you the problem in action. I am doing a calculation, clicking a link, and then clicking back. 
https://github.com/user-attachments/assets/574703c1-3420-4e79-84ac-2e8763cee070

This is the header animation in action, and fixed. It shows the shrinking heigh animation because that doesn't affect scroll seemingly. But I had to remove the height growth animation to prevent the aforementioned issues. 
https://github.com/user-attachments/assets/303f736e-b486-45ab-b2d0-c9e03f8c29ca
